File: scripts/runtime-usb-smoke.py
#!/usr/bin/env python3
"""Send a framed Studio request hex over serial and print the deframed reply hex."""

import sys
import time

import serial

SOF, ESC, EOF = 0xAB, 0xAC, 0xAD


def deframe(buf: bytes) -> bytes | None:
    mode = "idle"
    out = bytearray()
    for b in buf:
        if mode == "idle":
            if b == SOF:
                mode = "data"
                out = bytearray()
            continue
        if mode == "esc":
            out.append(b)
            mode = "data"
            continue
        if b == ESC:
            mode = "esc"
            continue
        if b == SOF:
            out = bytearray()
            continue
        if b == EOF:
            return bytes(out)
        out.append(b)
    return None


def main() -> int:
    if len(sys.argv) < 3:
        print("usage: runtime-usb-smoke.py PORT FRAMED_HEX", file=sys.stderr)
        return 2
    port, framed_hex = sys.argv[1], sys.argv[2]
    payload = bytes.fromhex(framed_hex)
    ser = serial.Serial(port, 115200, timeout=0.2)
    try:
        ser.reset_input_buffer()
        ser.write(payload)
        ser.flush()
        deadline = time.time() + 3.0
        buf = bytearray()
        while time.time() < deadline:
            chunk = ser.read(1024)
            if chunk:
                buf.extend(chunk)
                frame = deframe(bytes(buf))
                if frame:
                    print(frame.hex())
                    return 0
            else:
                time.sleep(0.05)
    finally:
        ser.close()
    print("no framed reply", file=sys.stderr)
    return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/runtime-usb-smoke.mjs
#!/usr/bin/env node
/**
 * USB smoke test for ZMK Next Runtime Config over Studio serial.
 *
 *   node scripts/runtime-usb-smoke.mjs            # print framed GetRuntimeCapabilities hex
 *   python3 scripts/runtime-usb-smoke.py /dev/cu.usbmodem*
 *
 * Does not flash. Safe to run against a test image. Uses the same framing and
 * protobuf codec as the configurator.
 */
import { spawnSync } from "node:child_process";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { frameBytes, encodeRuntimeRequestForTest, parseResponse } from "../src/connection/studio.js";

const root = join(dirname(fileURLToPath(import.meta.url)), "..");
const request = encodeRuntimeRequestForTest(1, 2);
const framed = frameBytes(request);
const hex = Buffer.from(framed).toString("hex");

if (process.argv.includes("--hex")) {
  process.stdout.write(hex);
  process.exit(0);
}

const port = process.argv[2];
if (!port) {
  console.log("framed GetRuntimeCapabilities:", hex);
  console.log("pass a serial port to send it: node scripts/runtime-usb-smoke.mjs /dev/cu.usbmodemXXXX");
  process.exit(0);
}

const py = spawnSync(
  process.env.PYTHON || `${process.env.HOME}/zmk-venv/bin/python`,
  [join(root, "scripts/runtime-usb-smoke.py"), port, hex],
  { encoding: "utf8" }
);
if (py.stdout) process.stdout.write(py.stdout);
if (py.stderr) process.stderr.write(py.stderr);
if (py.status !== 0) process.exit(py.status || 1);
const replyHex = py.stdout.trim().split("\n").at(-1) || "";
if (!/^[0-9a-f]+$/i.test(replyHex)) {
  console.error("no hex reply from serial");
  process.exit(1);
}
const reply = Buffer.from(replyHex, "hex");
const parsed = parseResponse(reply);
const caps = parsed.runtimeConfig?.capabilities;
if (!caps) {
  console.error("reply was not Runtime Config capabilities", JSON.stringify(parsed, null, 2));
  process.exit(1);
}
console.log(
  JSON.stringify(
    {
      protocolVersion: caps.protocolVersion,
      persistenceSchemaVersion: caps.persistenceSchemaVersion,
      selectedPositionCount: caps.selectedPositionCount,
      selectedToStockPositions: caps.selectedToStockPositions,
      supportedObjectTypes: caps.supportedObjectTypes,
      limits: caps.limits,
    },
    null,
    2
  )
);
if (caps.protocolVersion !== 1 || !caps.selectedToStockPositions?.length) {
  process.exit(1);
}
